# Sampling Assignment
### Name: Harshita Pandey
### Roll No: 102003235
### Group No: CO10


## Importing the dataset

In [30]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [21]:
df = pd.read_csv('Creditcard_data.csv')

In [22]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      1  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [23]:
df.shape

(772, 31)

In [24]:
freq = df["Class"].value_counts(normalize=True)
print(freq)
print(freq*100)

0    0.988342
1    0.011658
Name: Class, dtype: float64
0    98.834197
1     1.165803
Name: Class, dtype: float64


The given dataset is highly imbalanced. From the above calculations, we find out that the data is 98.83% imabalanced.

Since the dataset is highly imabalanced, we use the SMOTE or the Synthetic Minority Oversampling Technique

In [25]:
pip install imblearn

Note: you may need to restart the kernel to use updated packages.


In [26]:
from imblearn.over_sampling import SMOTE

In [61]:
# Separate input features and target variable
x = df.drop('Class', axis=1)
y = df['Class']

In [62]:
# Split dataset into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [63]:
# Apply SMOTE to training set only
smote = SMOTE(random_state=42)
x_train_resampled, y_train_resampled = smote.fit_resample(x_train, y_train)

In [64]:
oversample = SMOTE()
x, y = oversample.fit_resample(x, y)

In [65]:
y.value_counts()

0    763
1    763
Name: Class, dtype: int64

In [66]:
x_dummy, x_test, y_dummy, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [67]:
df2 = pd.concat([x, y], axis=1)
df2

Time        V1        V2        V3        V4        V5        V6  \
0        0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1        0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2        1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3        1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4        2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...    ...       ...       ...       ...       ...       ...       ...   
1521   145  0.312731 -0.494345  0.800036  0.638688  0.721476  0.103967   
1522   145  0.199690  0.518887  0.419685  0.152560  0.736859  0.213810   
1523   411 -0.700235  0.432964  1.447839  0.167270  0.933755 -0.638823   
1524   483 -1.977535  0.843877  0.022000  1.282213  0.319618 -0.764473   
1525   565  0.664007 -0.156439  0.700550  0.772529  0.156021 -0.764685   

            V7        V8        V9  ...       V21       V22       V23  \
0     0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1    -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2     0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3     0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4     0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...        ...       ...       ...  ...       ...       ...       ...   
1521 -0.638202  0.278761  0.078949  ... -0.047267 -0.133609  0.155413   
1522  0.253991  0.133359 -0.137987  ... -0.139705 -0.410902  0.145401   
1523  0.746457 -0.089232 -0.137780  ...  0.021967  0.130480 -0.164927   
1524 -0.899957  0.661749 -1.140196  ...  0.110699 -0.126643 -0.346578   
1525 -0.307415 -0.028331  0.186531  ... -0.157609 -0.453301  0.158671   

           V24       V25       V26       V27       V28      Amount  Class  
0     0.066928  0.128539 -0.189115  0.133558 -0.021053  149.620000      0  
1    -0.339846  0.167170  0.125895 -0.008983  0.014724    2.690000      1  
2    -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.660000      0  
3    -1.175575  0.647376 -0.221929  0.062723  0.061458  123.500000      0  
4     0.141267 -0.206010  0.502292  0.219422  0.215153   69.990000      0  
...        ...       ...       ...       ...       ...         ...    ...  
1521 -0.482288  0.060739  0.328332 -0.040828 -0.036437    2.362299      1  
1522 -1.248407 -1.213475  0.081109  0.204586  0.205860    1.181825      1  
1523 -0.012580 -0.013863 -0.333450 -0.027966 -0.044793    0.997725      1  
1524 -0.005176 -0.176934  0.214295 -0.092585 -0.283471    0.583448      1  
1525  0.124426  0.140865  0.234083 -0.040854 -0.005864    1.328265      1  

[1526 rows x 31 columns]

## Generating samples

In [68]:
z = 1.96 
p = 0.5
e = 0.05 

In [69]:
from sklearn.preprocessing import normalize
from imblearn.over_sampling import RandomOverSampler
import math
import random
from sklearn.cluster import KMeans

#### 1. Simple  Random Sampling

In [70]:
ssize_srs = (z*z*p*(1-p))/(e*e)
ssize_srs = math.floor(ssize_srs)
print(ssize_srs)
def random(df, ssize_srs):
    random_sample = df.sample(n = ssize_srs)
    return(random_sample)
Sample = random(df2, ssize_srs)

x_ssize_srs = Sample.drop('Class', axis=1) #X_train
y_ssize_srs = Sample['Class'] #y_train
     

384


#### 2. Systematic Sampling

In [77]:
t = 3 #interval
ssize_systematic = df2.iloc[::interval]
ssize_systematic

x_ssize_systematic = ssize_systematic.drop('Class', axis=1) #X_train
y_ssize_systematic = ssize_systematic['Class'] #y_train

#### 3. Stratified Sampling

In [78]:
s = 2 
ssize_stratified = (z*z*p*(1-p))/(e*e/s*s)
ssize_stratified = math.floor(ssize_srs)
print(ssize_stratified)
stratified = df2.groupby('Class')
stratified_df3 = stratified.apply(lambda x: x.sample(ssize_stratified))
stratified_df3
x_ssize_stratified = stratified_df3.drop('Class', axis=1) #X_train
y_ssize_stratified = stratified_df3['Class'] #y_train

384


#### 4. Cluster Sampling

In [81]:
kmeans = KMeans(n_clusters=10, random_state=42).fit(df2)
ssize_clustered = kmeans.labels_
cluster = [1,3,5,6,8]
cluster_series = pd.Series(ssize_clustered)
cluster_df = df2[cluster_series.isin(cluster)]
x_ssize_clustered = cluster_df.drop('Class', axis=1) #X_train
y_ssize_clustered = cluster_df['Class'] #y_train

#### 5. Snowball Sampling

In [114]:
def snowball(df, seed, n):
    snowball_sample = pd.DataFrame(columns=df.columns)
    snowball_sample = snowball_sample.append(df.loc[seed])
    current_sample = [seed]
    for i in range(n):
        rand_index = np.random.choice(len(current_sample))
        rand_observation = current_sample[rand_index]
        neighbors = df.loc[df['neighbor'] == rand_observation]
        new_neighbors = neighbors[~neighbors.index.isin(snowball_sample.index)]
        snowball_sample = snowball_sample.append(new_neighbors)
        current_sample.extend(new_neighbors.index.tolist())
    return snowball_sample
x_ssize_snowball = Sample.drop('Class', axis=1) #X_train
y_ssize_snowball = Sample['Class'] #y_train

## Training the models

In [102]:
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import BaggingClassifier

#### 1. Random Forest Classifier

In [116]:
def Random_forest_classifier(x_train, y_train, x_test, y_test):
  clf = RandomForestClassifier(n_estimators = 10)
  clf.fit(x_train, y_train)
  y_pred = clf.predict(x_test)
  return accuracy_score(y_test,y_pred)

#### 2. Bagging Classifier

In [104]:
def Bagging_classifier(x_train, y_train, x_test, y_test):
    base_classifier = DecisionTreeClassifier()
    bagging_classifier = BaggingClassifier(base_estimator=base_classifier, n_estimators=10)
    bagging_classifier.fit(x_train, y_train)
    y_pred = bagging_classifier.predict(x_test)
    return accuracy_score(y_test, y_pred)

#### 3. Decision Tree Classifier

In [105]:
def Decision_tree_classifier(x_train, y_train, x_test, y_test):
  tree = DecisionTreeClassifier(max_depth = 2).fit(x_train, y_train)
  tree_pred = tree.predict(x_test)
  return accuracy_score(y_test,tree_pred)

#### 4. Logistic Regression

In [106]:
def Logistic_regression(x_train, y_train, x_test, y_test):
  clf = LogisticRegression()
  clf.fit(x_train, y_train)
  y_pred = clf.predict(x_test)
  return accuracy_score(y_test,y_pred)

#### 5. K Neighbors Classifier

In [117]:
def KNN(x_train, y_train, x_test, y_test):
  knn = KNeighborsClassifier(n_neighbors = 10).fit(x_train, y_train)
  knn_pred = knn.predict(x_test)
  return accuracy_score(y_test,knn_pred)

## Calculating the accuracy for each sample type using each model

In [133]:
simple_random_Random_forest = Random_forest_classifier(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_bagging_classifier = Bagging_classifier(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_Decision_tree_classifier = Decision_tree_classifier(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_logistic = Logistic(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_KNN = KNN(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_accuracy = [simple_random_Random_forest*100,simple_random_bagging_classifier*100,simple_random_Decision_tree_classifier*100,simple_random_logistic*100,simple_random_KNN*100]
simple_random_accuracy

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[99.0079365079365,
 96.42857142857143,
 88.29365079365078,
 90.47619047619048,
 73.01587301587301]

In [134]:
systematic_Random_forest = Random_forest_classifier(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_bagging_classifier = Bagging_classifier(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_Decision_tree_classifier = Decision_tree_classifier(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_logistic = Logistic(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_KNN = KNN(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_accuracy = [systematic_Random_forest*100,systematic_bagging_classifier*100,systematic_Decision_tree_classifier*100,systematic_logistic*100,systematic_KNN*100]
systematic_accuracy


/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[99.20634920634922,
 97.02380952380952,
 89.0873015873016,
 89.68253968253968,
 77.38095238095238]

In [135]:
stratified_Random_forest = Random_forest_classifier(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_bagging_classifier = Bagging_classifier(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_Decision_tree_classifier = Decision_tree_classifier(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_logistic = Logistic(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_KNN = KNN(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_accuracy = [stratified_Random_forest*100,stratified_bagging_classifier*100,stratified_Decision_tree_classifier*100,stratified_logistic*100,stratified_KNN*100]
stratified_accuracy

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[99.40476190476191,
 98.80952380952381,
 89.48412698412699,
 92.26190476190477,
 80.75396825396825]

In [136]:
clustered_Random_forest = Random_forest_classifier(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_bagging_classifier = Bagging_classifier(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_Decision_tree_classifier = Decision_tree_classifier(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_logistic = Logistic(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_KNN = KNN(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_accuracy = [clustered_Random_forest*100,clustered_bagging_classifier*100,clustered_Decision_tree_classifier*100,clustered_logistic*100,clustered_KNN*100]
clustered_accuracy
     

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[98.80952380952381,
 99.40476190476191,
 89.0873015873016,
 89.0873015873016,
 83.73015873015873]

In [137]:
snowball_Random_forest = Random_forest_classifier(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_bagging_classifier = Bagging_classifier(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_Decision_tree_classifier = Decision_tree_classifier(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_logistic = Logistic(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_KNN = KNN(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_accuracy = [snowball_Random_forest*100,snowball_bagging_classifier*100,snowball_Decision_tree_classifier*100,snowball_logistic*100,snowball_KNN*100]
snowball_accuracy

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[98.21428571428571,
 97.02380952380952,
 88.29365079365078,
 90.47619047619048,
 73.01587301587301]

## Combining the results

In [148]:
final_accuracy = {
    'ML Model' : ['Random Forest','Bagging','Decision Tree','Logistic Regression','KNN'],
    'Simple Random Sampling(%age)' : simple_random_accuracy,
    'Stratified Sampling(%age)' : stratified_accuracy,
    'Cluster Sampling(%age)' : clustered_accuracy,
    'Systematic Sampling(%age)' : systematic_accuracy,
    'Snowball Sampling(%age)' : snowball_accuracy
}

final_accuracy_df = pd.DataFrame(final_accuracy)
final_accuracy_df

ML Model  Simple Random Sampling(%age)  \
0        Random Forest                     99.007937   
1              Bagging                     96.428571   
2        Decision Tree                     88.293651   
3  Logistic Regression                     90.476190   
4                  KNN                     73.015873   

   Stratified Sampling(%age)  Cluster Sampling(%age)  \
0                  99.404762               98.809524   
1                  98.809524               99.404762   
2                  89.484127               89.087302   
3                  92.261905               89.087302   
4                  80.753968               83.730159   

   Systematic Sampling(%age)  Snowball Sampling(%age)  
0                  99.206349                98.214286  
1                  97.023810                97.023810  
2                  89.087302                88.293651  
3                  89.682540                90.476190  
4                  77.380952                73.015873